### PR TITLE
Password show/hide functionality

### PR DESCRIPTION
**Pull request by hannah-whelan**

### CODE REVIEW

File: src/app/components/handlers/usePasswordToggle.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    >
+      <path d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM144 256a144 144 0 1 1 288 0 144 144 0 1 1 -288 0zm144-64c0 35.3-28.7 64-64 64c-7.1 0-13.9-1.2-20.3-3.3c-5.5-1.8-11.9 1.6-11.7 7.4c.3 6.9 1.3 13.8 3.2 20.7c13.7 51.2 66.4 81.6 117.6 67.9s81.6-66.4 67.9-117.6c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3z" />
+    </svg>
+  );
+
+  const Icon = visible ? eyeSlashIcon : eyeIcon;
+
+  const InputType = visible ? "text" : "password";
+
+  return [InputType, Icon];
```
Comment (jmgq) [inline src/app/components/handlers/usePasswordToggle.tsx:35]: Rather than an array, this should probably return an object, and you could specify the type/interface for that object. That way you don't have to write ` as ["text" | "password", React.ReactNode];` when you call this function.

File: src/app/components/handlers/usePasswordToggle.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import React, { useState } from "react";
+
+const usePasswordToggle = () => {
```
Comment (jmgq) [inline src/app/components/handlers/usePasswordToggle.tsx:3]: Rather than this function, wouldn't it make more sense to have a `Password` component? I'm guessing all the passwords on the website have the same functionality. So if this assumption is correct, then the clean way to do this would be with a `Password` component.

Comment (hannah-whelan) [reply src/app/components/handlers/usePasswordToggle.tsx]: Not entirely sure what you mean here; there are multiple password components through the site, and we want all of them to be controlled by a show/hide functionality, but only a single button. For example on the My Account page 

Screenshot 2023-08-30 at 12 07 26

Comment (jmgq) [reply src/app/components/handlers/usePasswordToggle.tsx]: Thanks for clarifying! However, what I meant is that it looks like the inputs that contain the eye icon should probably be a component by itself. At the moment they all look very similar, like this:
```jsx
<div className="password-group">
    <Input
        id="password-input" autoComplete="current-password" type={PasswordInputType} name="password" placeholder="Password"
        onChange={(event: React.ChangeEvent<HTMLInputElement>) => setPassword(event.target.value)}
        invalid={!!errorMessage || (!validPassword && (logInAttempted))}
        aria-describedby="passwordValidationMessage"
        required
    />
    <span className="password-toggle-icon" id="password-toggle">{ToggleIcon}</span>
</div>
```

So I would have expected all that (plus the `const [visible, setVisible] = useState(false);` logic) to be encapsulated in its own `PasswordInput` component, if that makes sense.

I think that the problem is that the original code was already quite convoluted. So what I'm asking is to evaluate how hard would it be to do it properly, instead of adding more complexity on top of something that wasn't very well executed in the first place.

Comment (hannah-whelan) [reply src/app/components/handlers/usePasswordToggle.tsx]: There are I think 7 instances of the password input, with differences between them that would need to be combined. That would be one for another ticket to fix as it would take a bit longer to cleanup the old code, or we'd need to delay adding the show/hide functionality until after the registration feature deploy. Definitely value to reducing the complexity!

I think even making a component just for the inputs that contain the eye icon wouldn't eliminate the need for the usePasswordToggle() because the inputs that don't have the icon still need their type controlled by the InputType

File: src/app/components/pages/LogIn.tsx [hunk 3]
```diff
@@ -182,21 +184,19 @@ export const EmailPasswordInputs =({setEmail, setPassword, validEmail, validPass
...
+                passwordFieldType="Current"
+                isPasswordVisible={isPasswordVisible}
+                setIsPasswordVisible={setIsPasswordVisible} 
                 onChange={(event: React.ChangeEvent<HTMLInputElement>) => setPassword(event.target.value)}
                 invalid={!!errorMessage || (!validPassword && (logInAttempted))}
-                aria-describedby="passwordValidationMessage"
-                required
+                showToggleIcon={true}
             />
-            <FormFeedback id="passwordValidationMessage">
```
Comment (jmgq) [inline src/app/components/pages/LogIn.tsx:197]: This `FormFeedback` is gone. Was this intentional?

File: src/app/components/pages/LogIn.tsx [hunk 4]
```diff
@@ -182,21 +184,19 @@ export const EmailPasswordInputs =({setEmail, setPassword, validEmail, validPass
...
 
         <FormGroup className="mb-0">
-            {displayLabels && (<><Label htmlFor="password-input">Password</Label>
+            {displayLabels && (<><Label htmlFor="current-password">Password</Label>
             <span id={`password-help-tooltip`} className="icon-help ml-1" />
             <UncontrolledTooltip target={`password-help-tooltip`} placement="bottom">
             {PASSWORD_REQUIREMENTS}
             </UncontrolledTooltip></>)}
-            <Input
-                id="password-input" autoComplete="current-password" type="password" name="password" placeholder="Password"
```
Comment (jmgq) [inline src/app/components/pages/LogIn.tsx:191]: There are multiple parameters that were present before, but are missing now, such as `placeholder`, `required`, `aria-describedby`, etc.

File: src/app/components/pages/LogIn.tsx (hunk 3 above)
Comment (hannah-whelan) [reply src/app/components/pages/LogIn.tsx:197]: Yep that's intentional to fix a current bug I noticed. We changed the password requirements, but we've grandfathered them in, so currently if a user tries to login with an old password ("test1234" for example) it will allow them to login but it will also show the password feedback briefly, which looks awful.  So I removed the PW feedback - the PW requirements are in a tooltip on that page anyway, and they get an error back from the API if the PW is wrong.

File: src/scss/common/registration.scss [hunk 5]
```diff
@@ -12,3 +12,15 @@
     display: block;
   }
 }
+
+.password-group {
```
Comment (jmgq) [inline src/scss/common/registration.scss:16]: I think it may be worth moving these somewhere else, since they're not really only used during registration.

File: src/app/components/pages/LogIn.tsx (hunk 4 above)
Comment (hannah-whelan) [reply src/app/components/pages/LogIn.tsx:191]: required is included by default in the Password component so not necessary as a prop, aria-describedby is no longer needed (due to no feedback) but is still an optional prop for Password as it is used elsewhere. Placeholder change was intentional as it looks better, although I realised I also need to remove it from the email field on login page, so fixed that now.

File: src/app/components/elements/panels/UserPassword.tsx [hunk 6]
```diff
@@ -99,25 +103,33 @@ export const UserPassword = (
...
+                            </Label>
+                            <Password
+                                passwordFieldType="Confirm"
+                                isPasswordVisible={isPasswordVisible}
+                                setIsPasswordVisible={setIsPasswordVisible}
+                                invalid={!!currentPassword && !isNewPasswordConfirmed}
+                                onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
+                                setNewPasswordConfirm(e.target.value);
+                                setMyUser(
+                                Object.assign({}, myUser, {
```
Comment (jmgq) [inline src/app/components/elements/panels/UserPassword.tsx:119]: Does this help? https://code.visualstudio.com/updates/v1_49#_only-format-modified-text

The current indentation of the modified lines makes it hard to read.

File: src/app/components/elements/inputs/Password.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  invalid?: boolean;
+  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  onBlur?: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  onFocus?: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  showToggleIcon?: boolean;
+  defaultValue?: string;
+  ariadescribedby?: string;
+}
+
+const Password: React.FC<PasswordProps> = (props) => {
```
Comment (jmgq) [inline src/app/components/elements/inputs/Password.tsx:18]: Since this is a new class, it should have some unit tests.

File: src/app/components/elements/panels/UserPassword.tsx (hunk 6 above)
Comment (hannah-whelan) [reply src/app/components/elements/panels/UserPassword.tsx:119]: That will be helpful going forward! Although it does modify the entire section of code, not just the modified lines - so now when I commit it will change everything in the return for the component - I'll do in its own commit.

File: src/app/components/elements/inputs/Password.tsx [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    </svg>
+  );
+
+  const toggleIcon = isPasswordVisible ? eyeSlashIcon : eyeIcon;
+
+  const type = isPasswordVisible ? "text" : "password";
+
+  const name =
+    passwordFieldType === "New"
+      ? "new-password"
```
Comment (jmgq) [inline src/app/components/elements/inputs/Password.tsx:64]: Nested/Chained ternary operators are hard to read and prone to errors. This should be a switch statement.

File: src/app/components/elements/inputs/Password.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      height="1em"
+      viewBox="0 0 576 512"
+      style={{ fill: "#000000" }}
+      onClick={() => setIsPasswordVisible(true)}
+    >
+      <path d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM144 256a144 144 0 1 1 288 0 144 144 0 1 1 -288 0zm144-64c0 35.3-28.7 64-64 64c-7.1 0-13.9-1.2-20.3-3.3c-5.5-1.8-11.9 1.6-11.7 7.4c.3 6.9 1.3 13.8 3.2 20.7c13.7 51.2 66.4 81.6 117.6 67.9s81.6-66.4 67.9-117.6c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3z" />
+    </svg>
+  );
+
+  const toggleIcon = isPasswordVisible ? eyeSlashIcon : eyeIcon;
```
Comment (jmgq) [inline src/app/components/elements/inputs/Password.tsx:58]: This is debatable, but since `toggleIcon`, `eyeSlashIcon` and `eyeIcon` are React components, isn't the convention that they should start with an uppercase letter?

File: src/app/components/elements/inputs/Password.tsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+import { Input } from "reactstrap";
+
+interface PasswordProps {
+  isPasswordVisible: boolean;
+  setIsPasswordVisible: React.Dispatch<React.SetStateAction<boolean>>;
+  passwordFieldType: "New" | "Confirm" | "Current";
+  disabled?: boolean;
+  invalid?: boolean;
+  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  onBlur?: (e: React.ChangeEvent<HTMLInputElement>) => void;
```
Comment (jmgq) [inline src/app/components/elements/inputs/Password.tsx:11]: `FocusEventHandler`

File: src/app/components/elements/inputs/Password.tsx [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  onFocus?: (e: React.ChangeEvent<HTMLInputElement>) => void;
```
Comment (jmgq) [inline src/app/components/elements/inputs/Password.tsx:12]: `FocusEventHandler`

File: src/app/components/elements/inputs/Password.tsx [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    <div className="password-group">
+      <Input
+        id={name}
+        type={type}
+        name={name}
+        required
+        autoComplete={autoComplete}
+        disabled={disabled}
+        invalid={invalid}
+        onChange={(e) => {
```
Comment (jmgq) [inline src/app/components/elements/inputs/Password.tsx:82]: Something does not feel right here. You're treating `onChange`, `onBlur` and `onFocus` differently. Can't you just pass whatever the user provided (like you're doing with `onFocus`)?

File: src/app/components/elements/inputs/Password.tsx (hunk 9 above)
Comment (jmgq) [reply src/app/components/elements/inputs/Password.tsx:58]: Actually, ignore me, they're not components, they're elements, aren't they?

Comment (hannah-whelan) [reply src/app/components/elements/inputs/Password.tsx:58]: Since I'm not using them elsewhere I think they work fine just as JS variables. It actually breaks if I turn them into Components - functionality sort of works in that case but not on first click of the eyeIcon (and tests then fail of course). 

I could get rid of `toggleIcon` arguably, and just directly use `isPasswordVisible ? eyeSlashIcon : eyeIcon` in the return but I used it for readability.

File: src/app/components/elements/inputs/Password.tsx (hunk 7 above)
Comment (hannah-whelan) [reply src/app/components/elements/inputs/Password.tsx:18]: Test coverage on Password.tsx is showing at 95.45% due to it being used in integration test in My Account, Registration and Login pages which I thought would be sufficient?

File: src/app/components/elements/inputs/Password.tsx (hunk 10 above)
Comment (hannah-whelan) [reply src/app/components/elements/inputs/Password.tsx:11]: Oops!

File: src/app/components/elements/inputs/Password.tsx (hunk 12 above)
Comment (hannah-whelan) [reply src/app/components/elements/inputs/Password.tsx:82]: I thought `onBlur` and `onChange` needed the event in order to access e.target.value.  I did onBlur is slightly different to onChange just because onChange is always there but onBlur is an optional prop.

But actually after testing it looks like I can just write it without writing `(e) =>` Looks better now!

File: src/app/components/elements/inputs/Password.tsx [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    passwordFieldType,
+    disabled,
+    invalid,
+    onChange,
+    onBlur,
+    onFocus,
+    showToggleIcon,
+    defaultValue,
+    ariadescribedby,
+  } = props;
```
[the remark this answers is not in the document]
Comment (hannah-whelan) [reply src/app/components/elements/inputs/Password.tsx:31]: Yeah I know this route would be a bit less verbose but I quite like being able to see them all listed

File: src/app/components/elements/inputs/Password.tsx (hunk 7 above)
Comment (jmgq) [reply src/app/components/elements/inputs/Password.tsx:18]: Kind of, but that has some problems: the coverage is not that important, it's more important to test the different features and behaviours of your component. You may be testing some of that behaviour indirectly in other tests, but that's not clear. Also, as a rule of thumb, when you create a new component you should probably have a few tests for that component in isolation. There is a lot of logic inside this component, and since it does not have its own test, what would happen the next time someone wants to modify the behaviour of this component? How would anyone know where to add the tests for the new behaviour?

Comment (hannah-whelan) [reply src/app/components/elements/inputs/Password.tsx:18]: Makes sense, I've added a bunch of unit tests for Password in isolation

File: src/app/components/elements/inputs/PasswordInput.tsx [hunk 14]
```diff
@@ -32,6 +33,7 @@ export const PasswordInputs = ({
   setUnverifiedPassword,
   defaultPassword,
 }: PasswordInputProps) => {
+  const [PasswordInputType, ToggleIcon] = usePasswordToggle() as ["text" | "password", React.ReactNode];
```
Comment (KasimAliAND) [inline src/app/components/elements/inputs/PasswordInput.tsx]: Just a small one. Noticed "PasswordInputType, ToggleIcon" are not camel cased compared to other consts

Comment (hannah-whelan) [reply src/app/components/elements/inputs/PasswordInput.tsx]: That hook was removed from the code

File: src/app/components/elements/panels/UserPassword.tsx [hunk 15]
```diff
@@ -36,6 +37,7 @@ export const UserPassword = (
             setPasswordResetRequested(true);
         }
     };
+    const [PasswordInputType, ToggleIcon] = usePasswordToggle() as ["text" | "password", React.ReactNode];
```
Comment (KasimAliAND) [inline src/app/components/elements/panels/UserPassword.tsx]: Same comment as before. Constant isn't camel cased.

Comment (hannah-whelan) [reply src/app/components/elements/panels/UserPassword.tsx]: This is no longer in the code 🙂

File: src/test/pages/StudentRegistration.test.tsx [hunk 16]
```diff
@@ -140,5 +137,28 @@ describe("Student Registration", () => {
...
+  it("Clicking the show password icon will make the text of password field visible, and the icon change", async () => {
+    renderStudentRegistration();
+    checkPageTitle("Register as a student");
+    const formFields = getFormFields();
+    const passwordInput = formFields.password() as HTMLInputElement;
+    const confirmPasswordInput =
+      formFields.confirmPassword() as HTMLInputElement;
+    [passwordInput, confirmPasswordInput].forEach((input) => {
+      expect(input.type).toBe("password");
+    });
```
Comment (KasimAliAND) [inline src/test/pages/StudentRegistration.test.tsx:151]: Extract the repeated code and create a separate function. That we can then call.

File: src/test/pages/StudentRegistration.test.tsx [hunk 17]
```diff
@@ -140,5 +137,28 @@ describe("Student Registration", () => {
...
+    const showPasswordToggle = screen.queryByTestId("show-password-icon");
+    if (showPasswordToggle) {
+      await userEvent.click(showPasswordToggle);
+    } else {
+      fail("Show password icon not found");
+    }
+    [passwordInput, confirmPasswordInput].forEach((input) => {
+      expect(input.type).toBe("text");
+    });
```
Comment (KasimAliAND) [inline src/test/pages/StudentRegistration.test.tsx:160]: Same comment as before.

File: src/test/pages/StudentRegistration.test.tsx (hunk 16 above)
Comment (hannah-whelan) [reply src/test/pages/StudentRegistration.test.tsx]: Done 👍

File: src/test/components/Password.test.tsx [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+        showToggleIcon: true,
+      });
+
+      toggleIcon = getByTestId("show-password-icon");
+      expect(passwordInput).toHaveAttribute("type", "password");
+      fireEvent.click(toggleIcon);
+      expect(setIsPasswordVisibleMock).toHaveBeenCalledTimes(1);
+      expect(setIsPasswordVisibleMock).toHaveBeenCalledWith(true);
+
+      const { container: updatedContainer } = renderPassword({
```
Comment (jmgq) [inline src/test/components/Password.test.tsx:68]: This is a bit confusing, because you're essentially rendering a new Password component, both of them share the same `setIspasswordVisibleMock`, and you've done this as part of the same test. It would have been better to do this as a separate test. You can leave it as it is, but please take it into account for future tests.

File: src/test/components/Password.test.tsx [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+      (passwordFieldType) => {
+        const { container } = renderPassword({
+          passwordFieldType: passwordFieldType,
+        });
+
+        let expectedName;
+
+        switch (passwordFieldType) {
+          case "New":
+            expectedName = "new-password";
```
[the remark this answers is not in the document]
Comment (hannah-whelan) [reply src/test/components/Password.test.tsx]: Done!

File: src/test/components/Password.test.tsx (hunk 18 above)
Comment (hannah-whelan) [reply src/test/components/Password.test.tsx]: I went ahead and changed this into two separate tests for setting the password as visible / hidden

File: src/test/components/Password.test.tsx [hunk 20]
```diff
@@ -54,32 +58,32 @@ describe("Password Component", () => {
...
-    test("shows toggle icon and handles click, changing password input type", () => {
-      const { getByTestId } = renderPassword({
+    test("shows toggle icon and handles click, changing password to visible", () => {
+      const { getByTestId, container } = renderPassword({
         showToggleIcon: true,
       });
-
+      passwordInput = container.querySelector(
+        'input[name="new-password"]'
+      ) as HTMLInputElement;
```
[the remark this answers is not in the document]
Comment (hannah-whelan) [reply src/test/components/Password.test.tsx]: I've refactored to more closely match how I've done the tests in my other PR, which should be less repetitive

File: src/test/components/elements/inputs/Password.test.tsx [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+  describe("Rendering Styles", () => {
+    test("renders with invalid style when 'invalid' prop is true", () => {
+      setupTest({ invalid: true });
+      const passwordInput = findPasswordInput("New");
+      expect(passwordInput.classList.contains("is-invalid")).toBe(true);
+    });
+
+    test.each(passwordFieldTypes)(
+      "renders with correct name attribute for passwordFieldType: %s",
```
Comment (jmgq) [inline src/test/components/elements/inputs/Password.test.tsx:98]: The input is no longer a string, so this test title does not render as you'd expect. You can reword it to take both the type and the expected name into account.

File: src/test/components/elements/inputs/Password.test.tsx [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+      const hideIcon = findToggleIcon("hide");
+      fireEvent.click(hideIcon);
+      expect(setIsPasswordVisibleMock).toHaveBeenCalledTimes(1);
+      expect(setIsPasswordVisibleMock).toHaveBeenCalledWith(false);
+    });
+
+    test("does not show toggle icon when 'showToggleIcon' prop is not provided", () => {
+      setupTest();
+      const toggleIcon = screen.queryByTestId("show-password-icon");
+      expect(toggleIcon).toBe(null);
```
Comment (jmgq) [inline src/test/components/elements/inputs/Password.test.tsx:86]: Using the more specific assertions has the advantage of providing better error messages. For instance, in this case you could use `toBeNull()`.

File: src/test/components/elements/inputs/Password.test.tsx [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+    });
+  });
+
+  describe("Rendering Styles", () => {
+    test("renders with invalid style when 'invalid' prop is true", () => {
+      setupTest({ invalid: true });
+      const passwordInput = findPasswordInput("New");
+      expect(passwordInput.classList.contains("is-invalid")).toBe(true);
```
Comment (jmgq) [inline src/test/components/elements/inputs/Password.test.tsx:94]:
```typescript
expect(passwordInput).toHaveClass("is-invalid");
```

**Review by RobertSaylesAND — APPROVED** [review]
